Issue #1498 Add integration test for deployment of EAP applications

agajdosi opened the pull request:
This test checks if templates added by xpaas addon are available and if they can be deployed successfully.

hferentschik commented on test/integration/features/addon-xpaas.feature:
```diff
@@ -0,0 +1,34 @@
+     And status code of HTTP request to "<http-endpoint>" of service "<service-name>" in namespace "myproject" is equal to "200"
+     And executing "oc delete all --all" succeeds
+
+  Examples: Values to be used on next Minishift start
```
I guess the example description is not quite right

agajdosi replied:
thanks for a catch, missed it, fixing it now

hferentschik commented:
@agajdosi See also https://github.com/minishift/minishift/pull/1658

praveenkumar commented:
@agajdosi How much extra time does it take when running this specific test case, also are we planning to make it as part of each PR of this is something only going to run on nightly?

LalatenduMohanty commented:
retest this please

gbraad commented:
retest this please

amitkrout commented on test/integration/features/addon-xpaas.feature:
```diff
@@ -0,0 +1,34 @@
+
+  Scenario Outline: User deploys, checks out and deletes several templates from XpaaS imagestream
+   Given Minishift has state "Running"
+    When executing "oc new-app <template-name>" succeeds
```
I don't understand this, how ```<template-name>``` is being parsed ?

hferentschik replied:
It is a so-called Secario Outline - https://github.com/cucumber/cucumber/wiki/Scenario-Outlines.

A scenario is repeated based on some variables defines in the examples table.

hferentschik commented:
> also are we planning to make it as part of each PR of this is something only going to run on nightly?

I think it would be time to create this nightly build we have been talking about

praveenkumar commented:
> I think it would be time to create this nightly build we have been talking about

@hferentschik  yes, agree 👍

agajdosi commented:
@praveenkumar It takes quite a while, sometimes maybe 15+ minutes. I am +1 for nightly test as @hferentschik mentioned.

agajdosi commented:
I see that deployment of OpenJDK example does not pass the Central-CI and also now I am seeing it on my local workstation. This was not here before, investigating more.

agajdosi commented:
So OpenJDK is not stable, interesting is that it can pass 6 times in a row and then fail 4 times in a row. I will investigate more by manual checks, where I am still unlucky to reproduce. However for this issue I think I could just remove this one example and later focus on adding it when I will get the root of the issue.

agajdosi commented:
@hferentschik Suggested changes implemented. The test is not going to be run on pull request. Please take a look when you have some time.

hferentschik commented:
>  Suggested changes implemented. The test is not going to be run on pull request. Please take a look when you have some time.

Ok, I take a look at it first thing tomorrow.

LalatenduMohanty commented on Makefile:
```diff
@@ -198,12 +198,12 @@ test: vendor $(ADDON_ASSET_FILE)
 integration: $(MINISHIFT_BINARY)
 	mkdir -p $(INTEGRATION_TEST_DIR)
-	go test -timeout $(TIMEOUT) $(REPOPATH)/test/integration --tags=integration -v -args --test-dir $(INTEGRATION_TEST_DIR) --binary $(MINISHIFT_BINARY) --run-before-feature="$(RUN_BEFORE_FEATURE)" --tags basic $(GODOG_OPTS)
+	go test -timeout $(TIMEOUT) $(REPOPATH)/test/integration --tags=integration -v -args --test-dir $(INTEGRATION_TEST_DIR) --binary $(MINISHIFT_BINARY) --run-before-feature="$(RUN_BEFORE_FEATURE)" --tags basic "$(GODOG_OPTS)"
```
@agajdosi Any specific reason for this change? It was working fine previously, right?

agajdosi replied:
@LalatenduMohanty It was working, because we did not use multiple tags concatenated with multiple "&&" operands. In this case all the tags between "&&"s got ignored. Due to this I have put it into brackets, which prevents this issue.

agajdosi replied:
@LalatenduMohanty Another possibility would be to use just "positive" tags and keep the negative just for `cdk-only` or `minishift-only`. Maybe this would be more readable. WDYT?

hferentschik replied:
-1, afaict this breaks the use of GODOG_OPTS completely. I actually created an issue - https://github.com/minishift/minishift/issues/1679, but it turns out that the quotes cause the problem. As soon as you provide GODOG_OPTS things fall apart. For example:

    $ make integration GODOG_OPTS="--tags addon-xpaas"

    flag provided but not defined: -tags addon-xpaas
    Usage of /var/folders/39/70n784613gd370g63ggg62lh0000gn/T/    go-build708633840/github.com/minishift/minishift/test/    integration/_test/integration.test:
      -binary string
        	Path to minishift binary
      -definitions
    
      -format string
        	Sets which format godog will use (default "pretty")
      -minishift-args string
        	Arguments to pass to minishift
      -no-colors
        	Disable colors in godog output
      -paths string
        	 (default "./features")
      -run-before-feature string
        	Set of minishift commands to be executed before every feature.     Individual commands must be delimited by a semicolon.
      -stop-on-failure
        	Stop when failure is found
      -tags string
        	Tags for godog test
      -test-dir string
        	Path to the directory in which to execute the tests
      -test.bench regexp
        	run only benchmarks matching regexp
      -test.benchmem
        	print memory allocations for benchmarks
      -test.benchtime d
        	run each benchmark for duration d (default 1s)
      -test.blockprofile file
        	write a goroutine blocking profile to file
      -test.blockprofilerate rate
        	set blocking profile rate (see runtime.SetBlockProfileRate) (    default 1)
      -test.count n
        	run tests and benchmarks n times (default 1)
      -test.coverprofile file
        	write a coverage profile to file
      -test.cpu list
        	comma-separated list of cpu counts to run each test with
      -test.cpuprofile file
        	write a cpu profile to file
      -test.list regexp
        	list tests, examples, and benchmarch maching regexp then exit
      -test.memprofile file
        	write a memory profile to file
      -test.memprofilerate rate
        	set memory profiling rate (see runtime.MemProfileRate)
      -test.mutexprofile string
        	write a mutex contention profile to the named file after     execution
      -test.mutexprofilefraction int
        	if >= 0, calls runtime.SetMutexProfileFraction() (default 1)
      -test.outputdir dir
        	write profiles to dir
      -test.parallel n
        	run at most n tests in parallel (default 4)
      -test.run regexp
        	run only tests and examples matching regexp
      -test.short
        	run smaller test suite to save time
      -test.timeout d
        	panic test binary after duration d (0 means unlimited)
      -test.trace file
        	write an execution trace to file
      -test.v
        	verbose: print additional output
    exit status 2
    FAIL	github.com/minishift/minishift/test/integration	0.067s
    make: *** [integration] Error 1    

I think some work is needed here. Documentation-wise as well as how the integration targets work. Maybe we should actually re-open #1679.

agajdosi commented:
Just rebased on current master.

LalatenduMohanty commented on test/integration/features/addon-xpaas.feature:
```diff
@@ -0,0 +1,33 @@
+     And status code of HTTP request to "<http-endpoint>" of service "<service-name>" in namespace "myproject" is equal to "200"
+     And executing "oc delete all --all" succeeds
+
+  Examples: Template names and values to be used with each of them
```
Suggestion: `Required information to test the templates`

agajdosi replied:
ok

hferentschik commented on centos_ci.sh:
```diff
@@ -298,7 +298,7 @@ function build_and_test() {
   make prerelease synopsis_docs link_check_docs
   # Run integration test with 'kvm' driver
-  MINISHIFT_VM_DRIVER=kvm make integration_all
+  MINISHIFT_VM_DRIVER=kvm make integration GODOG_OPTS="--tags=~coolstore&&~addon-xpaas"
```
I recall an issue where we wanted to make sure that the target called by CI is a direct Makefile target and that we did not want to use custom properties within the CI script. This was to make it clear which features are run in CI jsut be looking at the Makefile. What happened to this issue?

agajdosi replied:
@hferentschik I just remember talk about integration_nightly/all, however the new target for CI seems reasonable for me. I will create it and will not use exclusion of tags, but instead will define the test directly tag by tag so it is easily understandable and does not face the issue with multiple ampersands. For now I will just mention in docs that defining multiple exclusive tags has issues and thus we recommend to define tags inclusively.

Possible follow up fix for making the tags more usable could be probably to move the tags from GODOG_OPTS to its own new parameter GODOG_TAGS.

hferentschik commented on test/integration/features/addon-xpaas.feature:
```diff
@@ -0,0 +1,33 @@
+@addon-xpaas @addon
```
+1 Test itself looks great.

hferentschik commented:
@agajdosi, regarding the Makefile, I think we are better off using target-specific variables for GODOG_OPTS. See also https://www.gnu.org/software/make/manual/html_node/Target_002dspecific.html. This allows having defaults for differentiating between _integration_ and _integration_all_ target, as well as giving the flexibility to run any combination one wants using GODOG_OPTS from the command line.

Have a  look at https://github.com/agajdosi/minishift/pull/3 and see whether this solves the "problem" which made you adding the quotes.

agajdosi commented:
@hferentschik Thank for code example on PR and link, missed the target specific variables before. This could be the way, but it still faces the problems I tried to fixed with quotes.

I have modified the code of test runner a little bit to see more info about and found out the ampersands needs to be escaped, otherwise Go interprets them as logical AND and thus all tags after `&&` are interpreted as new command and so they are not parsed as part of `--tags`:

```gherkin
[agajdosi@dhcp-24-156 minishift]$ make integration GODOG_OPTS="--tags=~coolstore\&\&~addon-xpaas"
go install -pkgdir=/home/agajdosi/golang/src/github.com/minishift/minishift/out/bindata -ldflags="-X github.com/minishift/minishift/pkg/version.minishiftVersion=1.8.0 -X github.
com/minishift/minishift/pkg/version.b2dIsoVersion=v1.2.0 -X github.com/minishift/minishift/pkg/version.centOsIsoVersion=v1.3.0 -X github.com/minishift/minishift/pkg/version.mini
kubeIsoVersion=v0.24.0 -X github.com/minishift/minishift/pkg/version.openshiftVersion=v3.6.0 -X github.com/minishift/minishift/pkg/version.commitSha=ba5931a7" ./cmd/minishift
mkdir -p /home/agajdosi/golang/src/github.com/minishift/minishift/out/integration-test
go test -timeout 3600s github.com/minishift/minishift/test/integration --tags=integration -v -args --test-dir /home/agajdosi/golang/src/github.com/minishift/minishift/out/integr
ation-test --binary /home/agajdosi/golang/bin/minishift \
--run-before-feature="" --tags=~coolstore\&\&~addon-xpaas
TAGS: ~coolstore&&~addon-xpaas
Test run using Boot2Docker iso image.
Keeping Minishift cache directory '/home/agajdosi/golang/src/github.com/minishift/minishift/out/integration-test/cache' for test run.
```

Will investigate try it out on Windows if we can use this way.

agajdosi commented:
@hferentschik But I am still not sure where I should go with make targets:

> I recall an issue where we wanted to make sure that the target called by CI is a direct Makefile target and that we did not want to use custom properties within the CI script. This was to make it clear which features are run in CI jsut be looking at the Makefile.

vs. ` MINISHIFT_VM_DRIVER=kvm make integration GODOG_OPTS="--tags=~coolstore&&~addon-xpaas"` in Makefile you have provided in the PR.

Are we ok with defining the CI test run in Makefile with new `integration_ci` target, or better to define it in centos_ci.sh?

hferentschik commented:
> Are we ok with defining the CI test run in Makefile with new integration_pr target

+1 _integration_pr_ or _instgration_ci_.

hferentschik commented:
@agajdosi Why do you need the '&'. Can you not use multiple --tag for AND logic and comma seperated for _OR_ logic -  https://github.com/cucumber/cucumber/wiki/Tags#logically-anding-and-oring-tags?

hferentschik commented:
BTW, the quotes are a problem for Make. If we in the end need quotes as part of the Go command we need to do some escaping/magic in the Makefile.

agajdosi commented:
> +1 integration_pr or instgration_ci.

ack, great!

@hferentschik I am not able to find the docs for it now, but godog has AND implemented with `&&`, OR with comma as Gherkin has. It might be related to the way how Go's flag package behaves, currently for us only the last `--tags=xyz` is being parsed and used.

> BTW, the quotes are a problem for Make. If we in the end need quotes as part of the Go command we need to do some escaping/magic in the Makefile.

Agree, I would like to not use them, the escaping plan is now without need for addition of any new quotes.

Thanks for consultation, will ping once it is implemented.

agajdosi commented:
@hferentschik Fixed now, have tested on Fedora and Win7 (cmd+PS), the escaping of ampersands works fine, no need for adding more quotes in this PR. I have not changed the old targets with target specific variables as it works fine, just split the lines for better readability.

For documentation I have just added a mention about fact that `integration_pr` is used for PR tests - it is enough I think as it is not meant for use of developers, but more for CI.

praveenkumar commented on centos_ci.sh:
```diff
@@ -298,7 +298,7 @@ function build_and_test() {
   make prerelease synopsis_docs link_check_docs
   # Run integration test with 'kvm' driver
-  MINISHIFT_VM_DRIVER=kvm make integration_all
+  MINISHIFT_VM_DRIVER=kvm make integration_pr
```
This sound like a plan but the problem is we still don't have a nightly setup for centos CI which means if we merge this PR then other tests will be ignored. I think we should first take care of nightly builds and then pr specific one. @minishift/minishift-dev any other suggestion?

agajdosi replied:
@praveenkumar If we merge this it will ignore the coolstore and this feature, nothing will change in coverage. But I agree, that in next sprint it would be good to work on creating nightly.

budhrg replied:
New target for running PR specific tests is the right direction since at some point we need to decide what tests need to be run in PR.

hferentschik replied:
It beats me why we cannot create this nightly build. This is nothing else than a new entry in https://github.com/minishift/minishift-ci-jobs/blob/master/minishift-ci-index.yaml with a cron based trigger. @budhrg or am I missing something. Why do we need to wait for another sprint to put this in place?

budhrg commented:
LGTM :+1:

hferentschik commented:
@agajdosi 

>  have not changed the old targets with target specific variables as it works fine, just split the lines for better readability.

Can you explain to me then, how I just run the xpaas features? Either to just verify your pull request w/o having to run all the other tests or even more important to develop my own new feature. In the later case, I will be iterating for a while around a single feature file and I want to run just this feature file until I am happy with it. How do I do that.

According to the docs, I can use GODOG_OPTS and hence I try:

    $ make integration GODOG_OPTS="--paths `pwd`/test/integration/features/addon-xpaas.feature"
    go test -timeout 3600s github.com/minishift/minishift/test/integration     --tags=integration -v -args --test-dir /Users/hardy/work/    developer-platform/cdk-3/go/src/github.com/minishift/minishift/out/    integration-test --binary /Users/hardy/work/developer-platform/cdk-3/go/    bin/minishift \
    	--run-before-feature="" --tags=basic --paths /Users/hardy/work/    developer-platform/cdk-3/go/src/github.com/minishift/minishift/test/    integration/features/addon-xpaas.feature
    Test run using Boot2Docker iso image.
    Log successfully started, logging into: /Users/hardy/work/    developer-platform/cdk-3/go/src/github.com/minishift/minishift/out/    test-results/integration_2017-11-14_15-02-31.log
    Running Integration test in: /Users/hardy/work/developer-platform/cdk-3/    go/src/github.com/minishift/minishift/out/integration-test
    Using binary: /Users/hardy/work/developer-platform/cdk-3/go/bin/minishift
    Feature: xpaas add-on
      Xpaas add-on imports xPaaS templates and imagestreams,
      which are then available in OpenShift to the user.
    Error closing the log: invalid argument
    
    No scenarios
    No steps
    222.138221ms
    testing: warning: no tests to run
    PASS
    ok  	github.com/minishift/minishift/test/integration	0.303s    

As you can see, no tests run. The problem is that '--tags basic' is always applied and no matter what I specify in _GODOG_OPTS_, it gets ANDed to the specified tags or paths. 

Using target-specific variables I can override the '--tags basic'.

agajdosi commented:
@hferentschik On Amit's PR which introduced the integration_all target there was an agreement (it was pulled) that `--tags=basic` and `--tags=~coolstore` will be used as it is now present. Reason was that it will make the use easier, if one for example wants to use just different `--format`. However if one specifies the `--tags` then the previous `--tags` is overwritten. In your case then your command should look like:

`make integration GODOG_OPTS="--paths `pwd`/test/integration/features/addon-xpaas.feature --tags=addon-xpaas"`

I have tested it now and it works for me (have modified the name of the file so I am sure it is not taken from original feature folder):

```gherkin
[agajdosi@dhcp-24-156 minishift]$ make integration GODOG_OPTS="--paths=/home/agajdosi/addon-xpaas.feature --tags=addon-xpaas"
go install -pkgdir=/home/agajdosi/golang/src/github.com/minishift/minishift/out/bindata -ldflags="-X github.com/minishift/minishift/pkg/version.minishiftVersion=1.8.0 -X github.
com/minishift/minishift/pkg/version.b2dIsoVersion=v1.2.0 -X github.com/minishift/minishift/pkg/version.centOsIsoVersion=v1.3.0 -X github.com/minishift/minishift/pkg/version.mini
kubeIsoVersion=v0.24.0 -X github.com/minishift/minishift/pkg/version.openshiftVersion=v3.6.0 -X github.com/minishift/minishift/pkg/version.commitSha=6df129a3" ./cmd/minishift
mkdir -p /home/agajdosi/golang/src/github.com/minishift/minishift/out/integration-test
go test -timeout 3600s github.com/minishift/minishift/test/integration --tags=integration -v -args --test-dir /home/agajdosi/golang/src/github.com/minishift/minishift/out/integr
ation-test --binary /home/agajdosi/golang/bin/minishift \
--run-before-feature="" --tags=basic --paths=/home/agajdosi/addon-xpaas.feature --tags=addon-xpaas
Test run using Boot2Docker iso image.
Keeping Minishift cache directory '/home/agajdosi/golang/src/github.com/minishift/minishift/out/integration-test/cache' for test run.
Log successfully started, logging into: /home/agajdosi/golang/src/github.com/minishift/minishift/out/test-results/integration_2017-11-14_15-35-11.log
Running Integration test in: /home/agajdosi/golang/src/github.com/minishift/minishift/out/integration-test
Using binary: /home/agajdosi/golang/bin/minishift
Feature: xpaas add-on at agajdosis home folder
  Xpaas add-on imports xPaaS templates and imagestreams,
  which are then available in OpenShift to the user.

  Scenario: User enables the xpaas add-on                   # /home/agajdosi/addon-xpaas.feature:6
    When executing "minishift addons enable xpaas" succeeds # integration_test.go:726 -> executingMinishiftCommandSucceedsOrFails
    Then stdout should contain "Add-on 'xpaas' enabled"     # integration_test.go:668 -> commandReturnShouldContain

  Scenario: User starts Minishift                          # /home/agajdosi/addon-xpaas.feature:10
    Given Minishift has state "Does Not Exist"             # integration_test.go:162 -> *Minishift.shouldHaveState
```

agajdosi commented:
@hferentschik Forgot to mention: I just wanted to not change what was agreed by a bigger discussion before. But I am totally fine with your proposal, it is more readable and understandable indeed. So are we going that way? I think it is not a big deal, actually, just tell me :)

hferentschik commented:
> On Amit's PR which introduced the integration_all target there was an agreement (it was pulled) that --tags=basic and --tags=~coolstore will be used as it is now present. Reason was that it will make the use easier, if one for example wants to use just different --format. 

But using target specific variables you can have both. You can have a reasonable default as well as the ability to slice and dice. 

> However if one specifies the --tags then the previous --tags is overwritten. In your case then your command should look like:
> make integration GODOG_OPTS="--paths `pwd`/test/integration/features/addon-xpaas.feature --tags=addon-xpaas"

And how do I know that? In this case it works because you specify the feature file as well as an additional tag. This is not intuitive. Arguably you have a similar problem when using target specific variables in the sense of that when you use it you also need to specify which tests you are running, but that's still better imo. An alternative is to split  GODOG_OPTS and also introduce GODOG_TESTS. The former is used for formatting parameters and the latter to control things like the formatter. This way one can have the best of two worlds. 

Yet another option is to also add _integration_single_ which forces you to specify GODOG_OPTS and can be used for running single features during development. 

In its current form there is a conflict/overlap between GODOG_OPTS and the --tags flag which is "hardcoded".

hferentschik commented:
> So are we going that way? 

That would be my choice as an immediate improvement. We could then investigate to split  GODOG_OPTS into  GODOG_OPTS and GODOG_TESTS and/or add another target. Either way, a target specific variable is more understandable imo.

agajdosi commented:
@hferentschik ack 👍

agajdosi commented:
@hferentschik Missed your above above comment, will add the target specific variable and we can see later if we want to do bigger refactor.

agajdosi commented:
@hferentschik Fixed now, please take a look when you have some free time :)

hferentschik commented on docs/source/contributing/developing.adoc:
```diff
@@ -221,12 +221,14 @@ $ make integration
 $ make integration_all
 ----
 
+NOTE: There is also a `make integration_pr` target which is being run as part of pull request testing.
```
I think @robin-owen tries to convert all NOTE, TIP, etc into the multi-line version.  

\[NOTE]
\====
\====

Best to align.

@robin-owen, why don't you schedule a meeting and give us a short heads up on the most common style violations we use?

agajdosi replied:
@hferentschik Aah, missed that, will then help @robin-owen with this file. Update soon.

+1 for the sync around best practices and common mistakes we do.

hferentschik replied:
@agajdosi - fyi pull request #1684. I guess in the end it does not matter whether you update this here. I take it that @robin-owen needs to rebase this pull request anyways an then he can take care of it then.

> +1 for the sync around best practices and common mistakes we do.

Right, at least the core devs should be aligned on this, otherwise, there will be inconsistencies in no time. 

@robin-owen I think we need two things here. First the additional section in the writing docs document which outlines some of these syles, eg multiline NOTE over NOTE:, use of '{project}' over Minishift, etc. Once this is in place we still should have a short sync were we together go through this list once. Just to make sure everyone read it. @robin-owen WDYT?

agajdosi commented on docs/source/contributing/developing.adoc:
```diff
@@ -282,6 +287,12 @@ Set `no-colors` to true to disable ansi colors of Godog's output.
 Definitions::
 Set `definitions` to true to print all available step definitions.
 
+[NOTE]
```
@hferentschik Based on our conversation around tags I have added this note, it is connected to addition of target specific variables in this PR.

hferentschik commented on docs/source/contributing/developing.adoc:
```diff
@@ -221,12 +221,17 @@ $ make integration
 $ make integration_all
 ----
 
+[NOTE]
```
Thanks. @robin-owen will have to rebase anyways, but at least you will not forget again :-)
